pkg/cloud/aws/actuators/machine/actuator_test.go

```diff
@@ -498,3 +506,377 @@ func TestImmutableStateChange(t *testing.T) {
+							PrivateIpAddress: aws.String("1.2.3.4"),
+							PublicIpAddress:  aws.String("5.6.7.8"),
+							EnaSupport:       aws.Bool(true),
+							EbsOptimized:     aws.Bool(true),
+						},
+					},
+				},
+			},
+		},
+		nil).Times(1)
+
+	return mockEC2
+}
+
+func TestIsNodeJoin(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name                      string
+		inputScope                *actuators.MachineScope
+		inputControlplaneMachines []*clusterv1.Machine
+		expectedIsNodeJoin        bool
+		expectedError             error
+	}{
+		{
+			name:                      "should always join worker machines",
+			inputScope:                getWorkerMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        true,
+			expectedError:             nil,
+		},
+		{
+			name:                      "should return error for random machines joining",
+			inputScope:                getRandoMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("Unknown value %q for label `set` on machine %q", "random", "random-0"),
+		},
+		{
+			name:                      "should return error when called with invalid scope",
+			inputScope:                getInvalidMachineScope(t),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("failed to create machine scope for machine %q in namespace %q: failed to generate new scope from nil cluster", "master-0", "awesome-ns"),
+		},
+		{
+			name:                      "should return error for no `set` label",
+			inputScope:                getNoLabelMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        false,
+			expectedError:             fmt.Errorf("Unknown value %q for label `set` on machine %q", "", "nolabel-0"),
+		},
+		{
+			name:                      "should not join first controlplane machine",
+			inputScope:                getControlplaneMachineScope(t, nil),
+			inputControlplaneMachines: nil,
+			expectedIsNodeJoin:        false,
+			expectedError:             nil,
+		},
+		{
+			name: "should not join controlplane machine when no controlplane machines exist",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesNotFound([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError:             nil,
+		},
+		{
+			name: "should return error when unable to verify controlplane machine existence",
+			inputScope: getControlplaneMachineScope(
+				t,
+				getMockEC2APIDescribeInstancesFail([]string{"master-0", "master-1"}, mockCtrl),
+			),
+			inputControlplaneMachines: getTestControlplaneMachines(),
+			expectedIsNodeJoin:        false,
+			expectedError: fmt.Errorf(
```

Comment:
I think testing the exact error message is quite brittle, what do you think about simply testing nil or not for the error?

Reply:
Yeah, I understand your point. But for this method, I feel strict unit tests should be OK/ preferred.
If the error changes in a future change, the unit test will also need to be updated, which IMO, should be good, letting the unit test grow with the code.